singer-commons/src/main/java/com/pinterest/singer/loggingaudit/client/AuditEventKafkaSender.java

```diff
@@ -0,0 +1,416 @@
+                "host=" + host, "stage=" + stage.toString());
+      } catch (Exception e) {
+        OpenTsdbMetricConverter
+            .incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_KAFKA_PARTITIONS_REFRESH_ERROR, 1,
+                "host=" + host, "stage=" + stage.toString());
+      }
+    }
+  }
+
+  /**
+   *  Sender dequeue LoggingAuditEvents and send out to Kafka cluster. If send out to one partition
+   *  fails, it will choose another partition. For each event, it will try at most
+   *  umOfPartitionsToTrySending times (default 3 times) before dropping this events. Metrics are
+   *  used to track the queue size and usuage, number of events sent out to Kafka successfully, and
+   *  the number of events dropped.
+   */
+  @Override
+  public void run() {
+    LoggingAuditEvent event = null;
+    ProducerRecord<byte[], byte[]> record;
+    byte[] value = null;
+
+    while (!cancelled.get()) {
+      try {
+        refreshPartitionIfNeeded();
+        event = queue.poll(DEQUEUE_WAIT_IN_SECONDS, TimeUnit.SECONDS);
+        if (event != null) {
+          try {
+            value = serializer.serialize(event);
+            record =
+                new ProducerRecord<>(this.topic, getAlternatePartition(partitionInfoList.size()),
+                    null, value);
+            kafkaProducer.send(record, new KafkaProducerCallback(event));
+          } catch (TException e) {
+            LOG.debug("[{}] failed to construct ProducerRecord because of serialization exception.",
+                Thread.currentThread().getName(), e);
+            OpenTsdbMetricConverter
+                .incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_SERIALIZATION_EXCEPTION, 1,
+                    "host=" + host, "stage=" + stage.toString(),
+                    "logName=" + event.getLoggingAuditHeaders().getLogName());
+            eventTriedCount.remove(event.getLoggingAuditHeaders());
+          }
+        }
+      } catch (InterruptedException e) {
+        LOG.warn("[{}] got interrupted when polling the queue and while loop is ended!",
+            Thread.currentThread().getName(), e);
+        OpenTsdbMetricConverter
+            .incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_DEQUEUE_INTERRUPTED_EXCEPTION, 1,
+                "host=" + host, "stage=" + stage.toString());
+        break;
+      } catch (Exception e) {
+        LOG.warn("Exit the while loop and finish the thread execution due to exception: ", e);
+        OpenTsdbMetricConverter.incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_EXCEPTION, 1,
+            "host=" + host, "stage=" + stage.toString());
+        break;
+      }
+    }
+  }
+
+
+  public class KafkaProducerCallback implements Callback {
+
+    private LoggingAuditEvent event;
+
+    public KafkaProducerCallback(LoggingAuditEvent event) {
+      this.event = event;
+    }
+
+    @Override
+    public void onCompletion(RecordMetadata recordMetadata, Exception e) {
```

Comment:
since inside this method, when exception is not null (send failed), we are doing a lot of checks. This part can be factored out as a separate method. I can make this change in the update of the PR